src/app-layer-parser.c

```diff
@@ -1051,6 +1060,18 @@ void AppLayerParserTransactionsCleanup(Flow *f, const uint8_t pkt_dir)
         i++;
     }
 
+    if (tofree) {
+        for (i = 0; i < tofree_nb; i++) {
+            p->StateTransactionFree(alstate, tofree[i]);
+        }
+        SCFree(tofree);
+    } else {
+        // we are oom, try to free all txs
+        for (i = min; i < min + total_txs; i++) {
+            p->StateTransactionFree(alstate, i);
+        }
+    }
+
```

Comment:
Considering that `tofree` tries to allocate space for the total of transactions, which comes from `state.tx_id`, which doesn't change when transactions are freed, and many times we only try to clean one tx at a time (this was the case I've observed for DNS, at least), if we couldn't allocate `tofree`, shouldn't we just try to free the transaction that was marked as to be cleaned?

Reply:
> if we couldn't allocate tofree, shouldn't we just try to free the transaction that was marked as to be cleaned

Nope, because the point of this is to not remove element from a list while iterating over it...

With my PR, if allocation fails, we free all the transitions, so if there is only one transaction like DNS (or any unidirectional always complete protocol), we free this one transaction